src/wireviz/wv_dataclasses.py

```diff
@@ -0,0 +1,815 @@
+                self.scale = "false"
+            elif self.width and self.height:
+                self.scale = "both"
+            else:
+                self.scale = "true"  # When only one dimension is specified.
+
+        if self.fixedsize:
+            # If only one dimension is specified, compute the other
+            # because Graphviz requires both when fixedsize=True.
+            if self.height:
+                if not self.width:
+                    self.width = self.height * aspect_ratio(self.src)
+            else:
+                if self.width:
+                    self.height = self.width / aspect_ratio(self.src)
+
+
+@dataclass
+class PinClass:
+    index: int
+    id: str
+    label: str
+    color: MultiColor
+    parent: str  # designator of parent connector
+    _num_connections = 0  # incremented in Connector.connect()
+    _anonymous: bool = False  # true for pins on autogenerated connectors
+    _simple: bool = False  # true for simple connector
+
+    def __str__(self):
+        snippets = [  # use str() for each in case they are int or other non-str
+            str(self.parent) if not self._anonymous else "",
+            str(self.id) if not self._anonymous and not self._simple else "",
+            str(self.label) if self.label else "",
+        ]
+        return ":".join([snip for snip in snippets if snip != ""])
+
+
+@dataclass
+class Component:
+    category: Optional[str] = None  # currently only used by cables, to define bundles
+    type: Union[MultilineHypertext, List[MultilineHypertext]] = None
+    subtype: Union[MultilineHypertext, List[MultilineHypertext]] = None
+
+    # part number
+    partnumbers: PartNumberInfo = None  # filled by fill_partnumbers()
+    # the following are provided for user convenience and should not be accessed later.
+    # their contents are loaded into partnumbers during the child class __post_init__()
+    pn: str = None
+    manufacturer: str = None
+    mpn: str = None
+    supplier: str = None
+    spn: str = None
+    # BOM info
+    qty: NumberAndUnit = NumberAndUnit(1, None)
+    amount: Optional[NumberAndUnit] = None
+    sum_amounts_in_bom: bool = True
+    ignore_in_bom: bool = False
+    bom_id: Optional[str] = None  # to be filled after harness is built
+
+    def fill_partnumbers(self):
+        partnos = [self.pn, self.manufacturer, self.mpn, self.supplier, self.spn]
+        partnos = [remove_links(entry) for entry in partnos]
+        partnos = tuple(partnos)
+        self.partnumbers = PartNumberInfo(*partnos)
+
+    def parse_number_and_unit(
+        self,
+        inp: Optional[Union[NumberAndUnit, float, int, str]],
+        default_unit: Optional[str] = None,
+    ) -> Optional[NumberAndUnit]:
+        if inp is None:
+            return None
+        elif isinstance(inp, NumberAndUnit):
+            return inp
+        elif isinstance(inp, float) or isinstance(inp, int):
+            return NumberAndUnit(float(inp), default_unit)
```

Comment:
Why is `inp` forced to a `float` here and also in the `str` case below? Why not keep it `int` when possible to avoid the `1.0 x Item` in the output that has been the more natural `1 x Item` before this PR?

Reply:
True. Can be reverted.

I believe the `try: number = float(number)` in the `elif isinstance(inp, str):` section was a primitive way to see whether, after splitting the string into `number` and `unit`, the `number` part was actually a number. This can be rewritten in a way that does not force the conversion. Even `_ = float(number)` would be enough, but I welcome a more elegant check.

Reply:
Fixed.